### ISSUE TITLE
Get rid of mypy

### ISSUE THREAD

**Issue by abergeron** [mila-udem/myia#48]
I want to discuss this, but I feel like mypy is causing more problems that it solves.

Its type inference is minimal at best, it gets confused the moment you do something that is slightly complex (like a decorator) and is generally unhelpful.  Moreover, I realized that I spend about 10%-20% of my time figuring out how to organize the code to make it happy.  This is not productive and is starting to be a drain on my time.

If we have a consensus on this, I would submit a PR to remove mypy from travis.  I don't know if we want to keep the annotations in place or not.  They can be useful for documentation at least.

**Comment by breuleux** [comment]
I tend to agree. I don't spend much time on it personally, since I only bother with the annotations once I'm satisfied with the code and the tests pass, but in my experience it hasn't really been useful to fix any bugs -- it's mostly just 10-15 minutes of jumping through hoops to tell the checker that everything works as intended.

I also feel that using mypy's system sort of defeats the point of using a dynamic language, without delivering the advantages of having a static type system ((good) inference, soundness, guarantees, enforcement, trust). In my opinion it's kind of the worst of both worlds. It's a type system that requires putting in some work, but that I can't bring myself to actually *trust*.
